File: bookmarker-java/src/main/java/com/sivalabs/bookmarker/config/security/TokenHelper.java
package com.sivalabs.bookmarker.config.security;

import com.sivalabs.bookmarker.config.BookmarkerProperties;
import com.sivalabs.bookmarker.config.TimeProvider;
import com.sivalabs.bookmarker.domain.entity.User;
import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Component;

import javax.servlet.http.HttpServletRequest;
import java.util.Date;

@Component
@RequiredArgsConstructor
public class TokenHelper {

    private final BookmarkerProperties bookmarkerProperties;
    private final TimeProvider timeProvider;

    private static final String AUDIENCE_WEB = "web";
    private SignatureAlgorithm SIGNATURE_ALGORITHM = SignatureAlgorithm.HS512;

    public String getUsernameFromToken(String token) {
        String username;
        try {
            final Claims claims = this.getAllClaimsFromToken(token);
            username = claims.getSubject();
        } catch (Exception e) {
            username = null;
        }
        return username;
    }

    private Date getIssuedAtDateFromToken(String token) {
        Date issueAt;
        try {
            final Claims claims = this.getAllClaimsFromToken(token);
            issueAt = claims.getIssuedAt();
        } catch (Exception e) {
            issueAt = null;
        }
        return issueAt;
    }

    public String refreshToken(String token) {
        String refreshedToken;
        Date a = timeProvider.now();
        try {
            final Claims claims = this.getAllClaimsFromToken(token);
            claims.setIssuedAt(a);
            refreshedToken = Jwts.builder()
                    .setClaims(claims)
                    .setExpiration(generateExpirationDate())
                    .signWith( SIGNATURE_ALGORITHM, bookmarkerProperties.getJwt().getSecret() )
                    .compact();
        } catch (Exception e) {
            refreshedToken = null;
        }
        return refreshedToken;
    }

    public String generateToken(String username) {
        return Jwts.builder()
                .setIssuer( bookmarkerProperties.getJwt().getIssuer() )
                .setSubject(username)
                .setAudience(AUDIENCE_WEB)
                .setIssuedAt(timeProvider.now())
                .setExpiration(generateExpirationDate())
                .signWith( SIGNATURE_ALGORITHM, bookmarkerProperties.getJwt().getSecret() )
                .compact();
    }

    private Claims getAllClaimsFromToken(String token) {
        Claims claims;
        try {
            claims = Jwts.parser()
                    .setSigningKey(bookmarkerProperties.getJwt().getSecret())
                    .parseClaimsJws(token)
                    .getBody();
        } catch (Exception e) {
            claims = null;
        }
        return claims;
    }

    private Date generateExpirationDate() {
        return new Date(timeProvider.now().getTime() + bookmarkerProperties.getJwt().getExpiresIn() * 1000);
    }

    public Boolean validateToken(String token, UserDetails userDetails) {
        final String username = getUsernameFromToken(token);
        // User user = (User) userDetails;
        // final Date created = getIssuedAtDateFromToken(token);
        return (username != null && username.equals(userDetails.getUsername()));
    }

    public String getToken( HttpServletRequest request ) {
        /**
         *  Getting the token from Authentication header
         *  e.g Bearer your_token
         */
        String authHeader = getAuthHeaderFromHeader( request );
        if ( authHeader != null && authHeader.startsWith("Bearer ")) {
            return authHeader.substring(7);
        }

        return null;
    }

    private String getAuthHeaderFromHeader(HttpServletRequest request) {
        return request.getHeader(bookmarkerProperties.getJwt().getHeader());
    }

}

File: backend/src/main/kotlin/com/sivalabs/bookmarker/security/TokenHelper.kt
package com.sivalabs.bookmarker.security

import com.sivalabs.bookmarker.config.BookmarkerProperties
import com.sivalabs.bookmarker.config.TimeProvider
import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.SignatureAlgorithm
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.stereotype.Component

import javax.servlet.http.HttpServletRequest
import java.util.Date

@Component
class TokenHelper {

    @Autowired
    private lateinit var bookmarkerProperties: BookmarkerProperties

    @Autowired
    private lateinit var timeProvider: TimeProvider

    private val signatureAlgorithm = SignatureAlgorithm.HS512

    private val audienceWeb = "web"

    fun getUsernameFromToken(token: String): String {
        return this.getAllClaimsFromToken(token).subject
    }

    fun refreshToken(token: String): String {
        val claims = this.getAllClaimsFromToken(token)
        claims.issuedAt = timeProvider.now()
        return Jwts.builder()
                .setClaims(claims)
                .setExpiration(generateExpirationDate())
                .signWith(signatureAlgorithm, bookmarkerProperties.jwt.secret)
                .compact()
    }

    fun generateToken(username: String): String {
        return Jwts.builder()
                .setIssuer(bookmarkerProperties.jwt.issuer)
                .setSubject(username)
                .setAudience(audienceWeb)
                .setIssuedAt(timeProvider.now())
                .setExpiration(generateExpirationDate())
                .signWith(signatureAlgorithm, bookmarkerProperties.jwt.secret)
                .compact()
    }

    private fun getAllClaimsFromToken(token: String): Claims {
        return Jwts.parser()
                .setSigningKey(bookmarkerProperties.jwt.secret)
                .parseClaimsJws(token)
                .body
    }

    private fun generateExpirationDate(): Date {
        return Date(timeProvider.now().time + bookmarkerProperties.jwt.expiresIn * 1000)
    }

    fun validateToken(token: String, userDetails: UserDetails): Boolean {
        val username = getUsernameFromToken(token)
        return username == userDetails.username
    }

    fun getToken(request: HttpServletRequest): String? {
        val authHeader = getAuthHeaderFromHeader(request)
        return if (authHeader != null && authHeader.startsWith("Bearer ")) {
            authHeader.substring(7)
        } else null
    }

    fun getAuthHeaderFromHeader(request: HttpServletRequest): String? {
        return request.getHeader(bookmarkerProperties.jwt.header)
    }
}
